File: 53.js
function minJumps(arr) {
  const n = arr.length;
  const graph = {};
  arr.forEach((v, i) => (graph[v] ??= []).push(i));
  const q = [[0, 0]];
  const visited = new Set([0]);

  while (q.length) {
    const [i, d] = q.shift();
    if (i === n - 1) return d;
    for (const j of [i - 1, i + 1, ...(graph[arr[i]] || [])]) {
      if (j >= 0 && j < n && !visited.has(j)) {
        visited.add(j);
        q.push([j, d + 1]);
      }
    }
    graph[arr[i]] = [];
  }
  return -1;
}

File: 53.py
from collections import deque, defaultdict
def minJumps(arr):
    n = len(arr)
    graph = defaultdict(list)
    for i, val in enumerate(arr):
        graph[val].append(i)
    q = deque([(0,0)])
    visited = set([0])
    while q:
        i, d = q.popleft()
        if i == n-1:
            return d
        for j in [i-1, i+1] + graph[arr[i]]:
            if 0 <= j < n and j not in visited:
                visited.add(j)
                q.append((j, d+1))
        graph[arr[i]].clear()
